File: anagrams-kotlin/src/main/kotlin/com/bloomberg/assignment/Trie.kt
package com.bloomberg.assignment

/**
 * A trie node, can have any number of words "terminating" at it.
 */
class Trie {

    val children : MutableMap<Char, Trie> = mutableMapOf()
    val words : MutableList<String> = mutableListOf()

    fun accept(path: TriePath) {
        if (path.isEmpty()) {
            words.add(path.word)
            return
        }
        val char = path.nextChar()
        val child = children.getOrElse(char!!, { Trie() })
        child.accept(path.stepOne())
        children[char] = child
    }

    fun collectAnagrams(allAnagrams: MutableList<Collection<String>>) {

        if (words.size > 1) {
            allAnagrams.add(words)
        }
        children.forEach { it.value.collectAnagrams(allAnagrams) }
    }
}

File: anagrams-python/trie.py
#!/usr/bin/python

class Trie:

    def __init__(self):
        self.children = {}
        self.words = []

    def accept(self, triePath):
        if triePath.isEmpty():
            self.words.append(triePath.word)
            return
        char = triePath.nextChar()
        child = None
        if char in self.children:
            child = self.children[char]
        else:
            child = Trie()
        child.accept(triePath.stepOne())
        self.children[char] = child


    def collectAnagrams(self, allAnagrams):
        if len(self.words) > 1:
            allAnagrams.append(self.words)
        for key, value in self.children.iteritems():
            value.collectAnagrams(allAnagrams)
